### PR TITLE
AC-553: The app crashes if numeric data is entered in the "Country" field in Register patient.

### PR DESCRIPTION
**Pull request by talha3111997**
## Description of what I changed

To prevent app from crashing if a user accidentally entered numeric data in the Country EditText in AddEditPatientActivity whose label is "Register Patient"

## Issue I worked on
JIRA Issue: https://issues.openmrs.org/browse/AC-553

Note: Actually there is one change i.e commit tab is showing 2 because it is considering "Merge remote-tracking branch 'upstream/master'" as second commit. (Sorry I am new to Git)

Yes my code works. I have also uploaded demo on JIRA Issue link

- [x] My pull request is **based on the latest changes** of the master branch.
- [ ] My pull request only contains ONE single commit
 (the number above, next to the 'Commits' tab is 1)

 - [ ] I have added tests to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)

- [ ]  All new and existing tests passed.

### CODE REVIEW

**Comment by talha3111997** [comment]
Kindly just merge my one commit i.e "To prevent app from crashing if numberic data is entered in the Country EditText in AddEditPatientActivity whose label is Register Patient".
Second commit is not actually a commit, it is added just mistakenly. 
Sorry for inconvenience.

**Comment by nishchalraj** [comment]
According to me, there should not be any entry for numeric data, just blocking entering the numbers will fix the issue.

**Comment by talha3111997** [comment]
> According to me, there should not be any entry for numeric data, just blocking entering the numbers will fix the issue.

But how one can restrict the user to not enter any numeric data. or you are suggesting drop down list for countries

**Comment by f4ww4z** [comment]
@talha3111997 Are you still working on this?

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java [hunk 1]
```diff
@@ -292,6 +293,12 @@ private Person createPerson() {
             lastNameTIL.setErrorEnabled(false);
         }
 
+
+
+
```
[the remark this answers is not in the document]
Comment (talha3111997) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: Alright :)

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java [hunk 2]
```diff
@@ -515,7 +523,12 @@ private void addSuggestionsToAutoCompleTextView() {
     }
 
     private void addSuggestionsToCities() {
+
         String country_name = edcountry.getText().toString();
+        if(country_name.matches(".*\\d.*")){
+           countryerror.setText("Country name cannot have any numbers");
+        return;
```
[the remark this answers is not in the document]
Comment (talha3111997) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: Thanks for pointing

**Comment by talha3111997** [comment]
> @talha3111997 Are you still working on this?
Sorry, I was stuck in some work. Surely, I will make the suggested changes in a short while.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java [hunk 3]
```diff
@@ -256,6 +256,7 @@ private Person createPerson() {
         String middleNameError = getString(R.string.midname_invalid_error);
         // Invalid family name
         String familyNameError = getString(R.string.lname_invalid_error);
+      //  String countryNameError = "Country name cannot contain symbols or numbers";
```
[the remark this answers is not in the document]
Comment (talha3111997) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: Probably, I tried that but was unable get the desired result. That's why I hard coded it.

Exactly, you are right I should have removed this "commented out line" before committing.
I will make the changes.

**Comment by talha3111997** [comment]
@f4ww4z I have done with it, removed extra spaces and made the indentation correct.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java [hunk 4]
```diff
@@ -306,6 +307,12 @@ private Person createPerson() {
             lnameerror.setText(familyNameError);
         }
 
+
```
Comment (Abdelaty) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: @talha3111997 Please use Code formatter in android studio, It's easier than make it by your hand line after line.

Comment (talha3111997) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: Alright, thanks for the suggestion 👍

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java [hunk 5]
```diff
@@ -516,6 +516,10 @@ private void addSuggestionsToAutoCompleTextView() {
 
     private void addSuggestionsToCities() {
         String country_name = edcountry.getText().toString();
+        if(country_name.matches(".*\\d.*")){
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: @talha3111997 Can you create a new method with boolean return value for this in `TextUtil` ?

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java [hunk 6]
```diff
@@ -516,6 +516,10 @@ private void addSuggestionsToAutoCompleTextView() {
...
+           countryerror.setText("Country name cannot have any numbers");
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: Always place messages like this in `strings.xml`.

Comment (talha3111997) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: alright :)

Comment (talha3111997) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: Alright, thanks for guiding

Comment (talha3111997) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: I have tried to follow your guidelines. Kindly have a look please. So, that I would be able to make a new pull request as soon as possible.

I have placed following line of code in `strings.xml`
`<string name="country_invalid_error">Country name cannot contain numbers</string>`

and place following code in AddEditPatientFragment.java 
`  if(TextUtils.isDigitsOnly(country_name)) {
            countryerror.setText(getString(R.string.country_invalid_error));
            return;
        }`

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java (hunk 5 above)
Comment (f4ww4z) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: @talha3111997 that's correct, yes.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java (hunk 6 above)
Comment (f4ww4z) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: Yes, that's the way.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java (hunk 5 above)
Comment (talha3111997) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: One thing I want to clarify is that do I need to make another boolean function in which `StringUtils.isNumeric()` function is to be used or just I have to replace `TextUtils` with `StringUtils` in `if` statement in the same function.
Thanks for guiding.

Comment (talha3111997) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: As StringUtils.isNumeric() is in StringUtils class from Apache commons-lang (third-party library). For that, relevant jar file is needed to add in our project classpath to import the correct class.
While in Android, we have TextUtils through which we can done our task (In my opinion). So, what's your opinion on this

Comment (f4ww4z) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: @talha3111997 I'm talking about the StringUtils class in our own codebase. See [this file](https://github.com/openmrs/openmrs-contrib-android-client/blob/master/openmrs-client/src/main/java/org/openmrs/mobile/utilities/StringUtils.java).

Comment (talha3111997) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: Alright, I understood :)

**Pull request closed without merging**